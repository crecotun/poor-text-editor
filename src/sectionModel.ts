import {
  SectionIdType,
  FormattingType,
  SectionType,
  TextPositionType,
} from './types'
import { Emitter, DefaultEvents, createNanoEvents } from 'nanoevents'

class SectionModel {
  public readonly id: SectionIdType
  private _text: string
  private _formatting: FormattingType[]
  emitter: Emitter<DefaultEvents>

  constructor({ id, text, formatting }: SectionType) {
    this.id = id
    this._text = text
    this._formatting = formatting || []
    this.emitter = createNanoEvents()
  }

  get text(): string {
    return this._text
  }

  get formatting(): FormattingType[] {
    return this._formatting
  }

  toString(): SectionType {
    return {
      id: this.id,
      text: this.text,
      formatting: this.formatting,
    }
  }

  applyBold(position: TextPositionType) {
    // apply bold
    // define if part of the range is already styled
    // if it's it, just extend this number
    // if the whole range is applied / remove bold
    // if the part of the range is within current styling remove part of the bold
    // split in several methods
    this.emitter.emit('model:change')
  }

  insertText(text: string, position: TextPositionType) {
    const preText = this.text.slice(0, position.from)
    const postText = this.text.slice(position.from)
    let newText = preText + text + postText
    this._text = newText

    // define if the text was set before, after ot inbetween applied styling
    // and recalculate their positions

    this.emitter.emit('model:change')

    return newText
  }

  deleteText(position: TextPositionType) {
    // deletes part of the text, from - to position
    this.emitter.emit('model:change')
  }
}

export { SectionModel }
