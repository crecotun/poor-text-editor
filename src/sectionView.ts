import { SectionModel } from './sectionModel'

class SectionView {
  private model: SectionModel
  private $el: Element

  constructor(model: SectionModel) {
    this.model = model
    this.$el = document.createElement('div')
  }

  render() {
    const elementFragment = document.createDocumentFragment()

    // make every char as seperate element to apply styling easily and detect cursor position
    this.model.text.split('').forEach((char) => {
      const span = document.createElement('span')
      span.textContent = char
      elementFragment.appendChild(span)
    })

    this.$el.innerHTML = ''
    this.$el.appendChild(elementFragment)

    return this.$el
  }
}

export { SectionView }
