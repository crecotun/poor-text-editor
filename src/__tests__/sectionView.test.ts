import { SectionModel } from '../sectionModel'
import {
  sectionWith1BoldStyling,
  sectionWithoutStyling,
  sectionWith2BoldStyling,
  sectionWithBoldAndItalicStyling,
} from '../mockData'
import { SectionView } from '../sectionView'

describe('Section View', () => {
  test('render text witouth styling', () => {
    const model = new SectionModel(sectionWithoutStyling)
    const view = new SectionView(model)

    expect(view.render()).toMatchSnapshot()
  })
  test('render text with just 1 bold tag', () => {
    const model = new SectionModel(sectionWith1BoldStyling)
    const view = new SectionView(model)

    expect(view.render()).toMatchSnapshot()
  })
  test('render text with just 2 bold tags', () => {
    const model = new SectionModel(sectionWith2BoldStyling)
    const view = new SectionView(model)

    expect(view.render()).toMatchSnapshot()
  })
  test('render text with bold and italic', () => {
    const model = new SectionModel(sectionWithBoldAndItalicStyling)
    const view = new SectionView(model)

    expect(view.render()).toMatchSnapshot()
  })
})
