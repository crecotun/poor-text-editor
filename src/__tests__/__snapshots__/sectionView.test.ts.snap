// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section View render text 1`] = `
<div>
  <span>
    W
  </span>
  <span>
    h
  </span>
  <span>
    e
  </span>
  <span>
    n
  </span>
  <span>
     
  </span>
  <span>
    M
  </span>
  <span>
    r
  </span>
  <span>
    .
  </span>
  <span>
     
  </span>
  <span>
    B
  </span>
  <span>
    i
  </span>
  <span>
    l
  </span>
  <span>
    b
  </span>
  <span>
    o
  </span>
  <span>
     
  </span>
  <span>
    B
  </span>
  <span>
    a
  </span>
  <span>
    g
  </span>
  <span>
    g
  </span>
  <span>
    i
  </span>
  <span>
    n
  </span>
  <span>
    s
  </span>
  <span>
     
  </span>
  <span>
    o
  </span>
  <span>
    f
  </span>
  <span>
     
  </span>
  <span>
    B
  </span>
  <span>
    a
  </span>
  <span>
    g
  </span>
  <span>
     
  </span>
  <span>
    E
  </span>
  <span>
    n
  </span>
  <span>
    d
  </span>
  <span>
     
  </span>
  <span>
    a
  </span>
  <span>
    n
  </span>
  <span>
    n
  </span>
  <span>
    o
  </span>
  <span>
    u
  </span>
  <span>
    n
  </span>
  <span>
    c
  </span>
  <span>
    e
  </span>
  <span>
    d
  </span>
  <span>
     
  </span>
  <span>
    t
  </span>
  <span>
    h
  </span>
  <span>
    a
  </span>
  <span>
    t
  </span>
  <span>
     
  </span>
  <span>
    h
  </span>
  <span>
    e
  </span>
  <span>
     
  </span>
  <span>
    w
  </span>
  <span>
    o
  </span>
  <span>
    u
  </span>
  <span>
    l
  </span>
  <span>
    d
  </span>
  <span>
     
  </span>
  <span>
    s
  </span>
  <span>
    h
  </span>
  <span>
    o
  </span>
  <span>
    r
  </span>
  <span>
    t
  </span>
  <span>
    l
  </span>
  <span>
    y
  </span>
  <span>
     
  </span>
  <span>
    b
  </span>
  <span>
    e
  </span>
  <span>
     
  </span>
  <span>
    c
  </span>
  <span>
    e
  </span>
  <span>
    l
  </span>
  <span>
    e
  </span>
  <span>
    b
  </span>
  <span>
    r
  </span>
  <span>
    a
  </span>
  <span>
    t
  </span>
  <span>
    i
  </span>
  <span>
    n
  </span>
  <span>
    g
  </span>
  <span>
     
  </span>
  <span>
    h
  </span>
  <span>
    i
  </span>
  <span>
    s
  </span>
  <span>
     
  </span>
  <span>
    e
  </span>
  <span>
    l
  </span>
  <span>
    e
  </span>
  <span>
    v
  </span>
  <span>
    e
  </span>
  <span>
    n
  </span>
  <span>
    t
  </span>
  <span>
    y
  </span>
  <span>
    -
  </span>
  <span>
    f
  </span>
  <span>
    i
  </span>
  <span>
    r
  </span>
  <span>
    s
  </span>
  <span>
    t
  </span>
  <span>
     
  </span>
  <span>
    b
  </span>
  <span>
    i
  </span>
  <span>
    r
  </span>
  <span>
    t
  </span>
  <span>
    h
  </span>
  <span>
    d
  </span>
  <span>
    a
  </span>
  <span>
    y
  </span>
  <span>
     
  </span>
  <span>
    w
  </span>
  <span>
    i
  </span>
  <span>
    t
  </span>
  <span>
    h
  </span>
  <span>
     
  </span>
  <span>
    a
  </span>
  <span>
     
  </span>
  <span>
    p
  </span>
  <span>
    a
  </span>
  <span>
    r
  </span>
  <span>
    t
  </span>
  <span>
    y
  </span>
  <span>
     
  </span>
  <span>
    o
  </span>
  <span>
    f
  </span>
  <span>
     
  </span>
  <span>
    s
  </span>
  <span>
    p
  </span>
  <span>
    e
  </span>
  <span>
    c
  </span>
  <span>
    i
  </span>
  <span>
    a
  </span>
  <span>
    l
  </span>
  <span>
     
  </span>
  <span>
    m
  </span>
  <span>
    a
  </span>
  <span>
    g
  </span>
  <span>
    n
  </span>
  <span>
    i
  </span>
  <span>
    f
  </span>
  <span>
    i
  </span>
  <span>
    c
  </span>
  <span>
    e
  </span>
  <span>
    n
  </span>
  <span>
    c
  </span>
  <span>
    e
  </span>
  <span>
    ,
  </span>
  <span>
     
  </span>
  <span>
    t
  </span>
  <span>
    h
  </span>
  <span>
    e
  </span>
  <span>
    r
  </span>
  <span>
    e
  </span>
  <span>
     
  </span>
  <span>
    w
  </span>
  <span>
    a
  </span>
  <span>
    s
  </span>
  <span>
     
  </span>
  <span>
    m
  </span>
  <span>
    u
  </span>
  <span>
    c
  </span>
  <span>
    h
  </span>
  <span>
     
  </span>
  <span>
    t
  </span>
  <span>
    a
  </span>
  <span>
    l
  </span>
  <span>
    k
  </span>
  <span>
     
  </span>
  <span>
    a
  </span>
  <span>
    n
  </span>
  <span>
    d
  </span>
  <span>
     
  </span>
  <span>
    e
  </span>
  <span>
    x
  </span>
  <span>
    c
  </span>
  <span>
    i
  </span>
  <span>
    t
  </span>
  <span>
    e
  </span>
  <span>
    m
  </span>
  <span>
    e
  </span>
  <span>
    n
  </span>
  <span>
    t
  </span>
  <span>
     
  </span>
  <span>
    i
  </span>
  <span>
    n
  </span>
  <span>
     
  </span>
  <span>
    H
  </span>
  <span>
    o
  </span>
  <span>
    b
  </span>
  <span>
    b
  </span>
  <span>
    i
  </span>
  <span>
    t
  </span>
  <span>
    o
  </span>
  <span>
    n
  </span>
  <span>
    .
  </span>
</div>
`;
